python/ray/ml/trainer.py

```diff
@@ -171,7 +188,32 @@ def preprocess_datasets(self) -> None:
         The transformed datasets will be set back in the ``self.datasets`` attribute
         of the Trainer to be used when overriding ``training_loop``.
         """
-        raise NotImplementedError
+        # Evaluate all datasets.
+        self.datasets = {k: d() if callable(d) else d for k, d in self.datasets.items()}
+
+        # Transform all datasets concurrently in remote tasks.
+        transform_task_dict = {}
+
+        transform_task = ray.remote(
+            lambda preprocessor, dataset: preprocessor.transform(dataset)
+        )
+
+        if self.preprocessor and not self._has_preprocessed_datasets:
+            train_dataset = self.datasets.get(TRAIN_DATASET_KEY, None)
+            if train_dataset and not self.preprocessor.check_is_fitted():
+                self.preprocessor.fit(train_dataset)
+
+            for key, dataset in self.datasets.items():
+                transform_task_dict[key] = transform_task.remote(
+                    self.preprocessor, dataset
+                )
+
+            ray.get(list(transform_task_dict.values()))
+
+            for key, transformed_dataset in transform_task_dict.items():
+                self.datasets[key] = ray.get(transformed_dataset)
+
+            self._has_preprocessed_datasets = True
```

Comment:
Actually, you're right I don't think this actually makes an impact. Since `preprocess_datasets` is only called in the Trainable, the datasets on the driver are not going to be mutated.

I will remove this!

Reply:
Removed!